wire-ios-data-model/Source/MLS/OneOnOne/OneOnOneResolver.swift

```diff
@@ -92,68 +91,151 @@ public final class OneOnOneResolver: OneOnOneResolverInterface {
+            throw OneOnOneResolverError.migratorNotFound
+        }
+
+        guard let mlsService else {
+            throw OneOnOneResolverError.migratorNotFound
+        }
+
+        let mlsGroupID = try await syncMLSConversationFromBackend(
+            userID: userID,
+            in: context
+        )
+
+        if await mlsService.conversationExists(groupID: mlsGroupID) {
+            return .noAction
+        }
+
+        guard let epoch = await fetchMLSConversationEpoch(mlsGroupID: mlsGroupID, in: context) else {
+            throw MigrateMLSOneOnOneConversationError.missingConversationEpoch
+        }
+
+        if epoch == 0 {
+            // migrate to a new conversation
+            try await migrator.migrateToMLS(
                 userID: userID,
+                mlsGroupID: mlsGroupID,
                 in: context
             )
+        } else {
+            // join existing conversation via external commit
+            try await mlsService.joinGroup(with: mlsGroupID)
+        }
 
-            return .migratedToMLSGroup(identifier: mlsGroupIdentifier)
+        return .migratedToMLSGroup(identifier: mlsGroupID)
+    }
 
-        case .proteus:
-            WireLogger.conversation.debug("should resolve to proteus 1-1 conversation")
-            return .noAction
+    private func syncMLSConversationFromBackend(
+        userID: QualifiedID,
+        in context: NSManagedObjectContext
+    ) async throws -> MLSGroupID {
+        var action = SyncMLSOneToOneConversationAction(
+            userID: userID.uuid,
+            domain: userID.domain
+        )
+
+        do {
+            return try await action.perform(in: context.notificationContext)
+        } catch {
+            throw MigrateMLSOneOnOneConversationError.failedToFetchConversation(error)
+        }
+    }
+
+    private func fetchMLSConversationEpoch(
+        mlsGroupID: MLSGroupID,
+        in context: NSManagedObjectContext
+    ) async -> UInt64? {
+        await context.perform {
+            let conversation = ZMConversation.fetch(with: mlsGroupID, in: context)
+            return conversation?.epoch
+        }
+    }
+
+    // MARK: Resolve - Proteus
+
+    private func resolveCommonUserProtocolProteus() -> OneOnOneConversationResolution {
+        WireLogger.conversation.debug("should resolve to proteus 1-1 conversation")
+        return .noAction
+    }
 
+    // MARK: Resolve - Mixed
+
+    private func resolveCommonUserProtocolMixed() -> OneOnOneConversationResolution {
         // This should never happen:
         // Users can only support proteus and mls protocols.
         // Mixed protocol is used by conversations to represent
         // the migration state when migrating from proteus to mls.
-        case .mixed:
-            assertionFailure("users should not have mixed protocol")
-            return .noAction
-        }
+        assertionFailure("users should not have mixed protocol")
+        return .noAction
     }
```

Comment:
in this case I would prefer not to have this in a separate method, I think it's clearer if the comment and the assertion are shown directly within the `mixed` case

Reply:
My idea was to have the same level of abstraction in the switch, but I can change it back :)

<img width="960" alt="Screenshot 2024-03-21 at 12 47 42" src="https://github.com/wireapp/wire-ios/assets/7196958/35afd45e-2aae-4c79-a260-22d3993a84ac">